src/domain.rs

```diff
@@ -494,3 +509,72 @@ fn parallel_fft_consistency() {
 
     test_consistency::<Bls12, _>(rng);
 }
+
+pub fn gpu_fft_supported(log_d: u32) -> ocl::Result<gpu::FFTKernel> {
+    use std::time::Instant;
+    use pairing::bls12_381::{Bls12};
+    use rand::{Rand};
+    let rng = &mut rand::thread_rng();
+
+    let worker = Worker::new();
+    let log_cpus = worker.log_num_cpus();
+
+    let mut kern = gpu::initialize(1 << log_d)?;
+
+    let log_d_test = 20;
+    let d_test = 1 << log_d_test;
+
+    let elems = (0..d_test).map(|_| Scalar::<Bls12>(Fr::rand(rng))).collect::<Vec<_>>();
+    let mut v1 = EvaluationDomain::from_coeffs(elems.clone()).unwrap();
+    let mut v2 = EvaluationDomain::from_coeffs(elems.clone()).unwrap();
+
+    let mut now = Instant::now();
+    bls12_gpu_fft(&mut kern, &mut v1.coeffs, &v1.omega, log_d_test)?;
+    let gpu_dur = now.elapsed().as_secs() * 1000 as u64 + now.elapsed().subsec_millis() as u64;
+
+    now = Instant::now();
+    if log_d_test <= log_cpus { serial_fft(&mut v2.coeffs, &v2.omega, log_d_test); }
+    else { parallel_fft(&mut v2.coeffs, &worker, &v2.omega, log_d_test, log_cpus); }
+    let cpu_dur = now.elapsed().as_secs() * 1000 as u64 + now.elapsed().subsec_millis() as u64;
+
+    if v1.coeffs == v2.coeffs && gpu_dur < cpu_dur { Ok(kern) }
+    else { Err(ocl::Error::from("GPU FFT is not working well on your machine!")) }
+}
+
+//#[test]
```

Comment:
why is this commented out?

Reply:
We couldn't test this function with `cargo test` as other tests were failing, as @nginnever mentioned.